### PR TITLE
Variable external density

### PR DESCRIPTION
**Pull request by stephankramer**
This allows for the external density under the viscous free surface to be space-varying. The option is changed from a simple real to a choice between constant, python or from_field.

### CODE REVIEW

File: main/Fluids.F90 [hunk 1]
```diff
@@ -865,11 +865,6 @@ SUBROUTINE FLUIDS()
           
        ! Call the modern and significantly less satanic version of study
        call write_diagnostics(state, current_time, dt, timestep)
-       ! Work out the domain volume by integrating the water depth function over the surface if using wetting and drying
```
Comment (cianwilson) [inline main/Fluids.F90]: Don't object to this being removed as it's a very random thing to be here but it seems orthogonal to the rest of the pull request.  Worth at least mentioning in the summary?

File: preprocessor/Boundary_Conditions_From_Options.F90 [hunk 2]
```diff
@@ -1187,6 +1193,26 @@ subroutine set_vector_boundary_conditions_values(state, field, bc_path, position
...
+           if (have_option(trim(bc_path_i)//"/external_density")) then
+              call get_boundary_condition(field, i+1, surface_mesh=surface_mesh, &
+                 surface_element_list=surface_element_list)
+              scalar_surface_field => extract_scalar_surface_field(field, bc_name, name="ExternalDensity")
+              if (have_option(trim(bc_path_i)//"/external_density/from_field")) then
+                ! from_field: The parent field contains the boundary values that you want to apply to surface_field.
+                call get_option(trim(bc_path_i)//"/external_density/from_field/parent_field_name", parent_field_name)
+                scalar_parent_field => extract_scalar_field(state, parent_field_name, stat)
+                if(stat /= 0) then
+                  FLExit("Could not extract parent field. Check options file?")
```
Comment (cianwilson) [inline preprocessor/Boundary_Conditions_From_Options.F90]: Perhaps worth adding to the error here "parent *scalar* field" as you've only checked for that (understandably) but the schema documentation (because its been reused from other purposes) is unclear about this.

**Comment by tmbgreaves** [comment]
Cian Wilson wrote:
> Buildbot?

minor glitch with buildbot at the moment (I've
misconfigured ssh and am off-site..) but will at
the latest sort it out in a couple of weeks when
I'm back on site again.

Tim

**Comment by drhodrid** [comment]
Green buildbot here: http://buildbot.ese.ic.ac.uk:8080/builders/variable-external-density

**Comment by stephankramer** [comment]
Rebased on master, implemented changes requested by @cianwilson. Have given buildbot one more spin with this
